crates/turbopack-ecmascript/src/chunk/placeable.rs

```diff
@@ -1,14 +1,47 @@
+use anyhow::Result;
 use turbo_tasks::Vc;
-use turbopack_core::{asset::Asset, chunk::ChunkableModule, module::Module};
+use turbo_tasks_fs::{FileJsonContent, FileSystemPath};
+use turbopack_core::{
+    asset::Asset,
+    chunk::ChunkableModule,
+    module::Module,
+    resolve::{find_context_file, package_json, FindContextFileResult},
+};
 
-use crate::references::{async_module::OptionAsyncModule, esm::EsmExports};
+use crate::references::{
+    async_module::OptionAsyncModule,
+    esm::{EsmExport, EsmExports},
+};
 
 #[turbo_tasks::value_trait]
 pub trait EcmascriptChunkPlaceable: ChunkableModule + Module + Asset {
     fn get_exports(self: Vc<Self>) -> Vc<EcmascriptExports>;
     fn get_async_module(self: Vc<Self>) -> Vc<OptionAsyncModule> {
         Vc::cell(None)
     }
+    fn is_marked_as_side_effect_free(self: Vc<Self>) -> Vc<bool> {
+        is_marked_as_side_effect_free(self.ident().path())
+    }
+}
+
+#[turbo_tasks::function]
+pub async fn is_marked_as_side_effect_free(path: Vc<FileSystemPath>) -> Result<Vc<bool>> {
+    let find_package_json: turbo_tasks::ReadRef<FindContextFileResult> =
+        find_context_file(path.parent(), package_json()).await?;
+
+    if let FindContextFileResult::Found(package_json, _) = *find_package_json {
+        if let FileJsonContent::Content(content) = &*package_json.read_json().await? {
+            if let Some(side_effects) = content.get("sideEffects") {
+                if let Some(side_effects) = side_effects.as_bool() {
+                    return Ok(Vc::cell(!side_effects));
+                } else {
+                    // TODO it might be a glob too, handle that case too
```

Comment:
should probably create a linear issue for all the todos and reference them in the comments